rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,108 @@
+
+There is one more case when a user can lock himself out:
+
+| initial state | change                                                                                    |
+| ------------- | ----------------------------------------------------------------------------------------- |
+| _admin_       | _admin_ loses acess to own account (eg. he stores credentials and recovery codes locally) |
+
+However, requiring a second admin to guard this situation is not good for user experience. Existing and new clusters will have to add a second _admin_.
+
+## Terminology
+
+- _admin_: local user with user/role editing capabilities
+- _SSO admin_: user with user/role editing capabilities added via SSO connector (eg. GitHub)
+
+## Why
+
+When a user locks themselves out, they are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
+
+Further, for our Cloud offering, we don't perform account resets as a policy. The only recourse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data loss.
+
+On-prem customers could work around this "locked out" state by running `tctl` locally on the auth server to create a new admin user.
+
+## Details
+
+> Context: SSO users
+>
+> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
+>
+> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
+> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
+> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
+> 4. SSO users can have full access.
+
+The solution here could be to introduce a new rule:
+
+- There should be at least _admin_
+
+This will ensure that:
+
+- Users won't be able to delete _admin_ user if there is only one _admin_. They would have to add a second one and then they can delete it.
+- Users won't be able to unassign the user/role editing capabilities from _admin_ user if there are only is only one _admin_. They would have to add a second one and then they can unassign.
+- Locking will be impossible on the SSO side since there will be always _admin_ on our side.
+
+Now we need to define what _admin_ means in our system. W can take 2 approaches:
+
+1. User with an immutable `editor` role is _admin_.
+2. User with a role that grants user/role editing capabilities is _admin_.
+
+From the user's perspective, the first approach is simpler. The error message saying: "You can't delete the last user with `editor` role." will be easier to understand than "You can't delete the last user with: list of role management capabilities".
```

Comment:
+1 for this - simple to explain and to implement. It's easy to have regressions in systems like this, which we are only likely to find out once someone gets locked out (again). It's much simpler to ensure that an immutable role performs whatever actions it needs to.

Reply:
The immutable roles would require users to modify their cluster: 

> complicated migration in case of updating cluster. It is still possible to delete/edit default roles so not all existing clusters will have default roles in the initial form. Maybe the `editor` role is used as a role with a different meaning than the initial one? We can't override existing roles. So the solution would be to add 3 new roles (instead of editing/re-adding: editor, auditor, access). This will ensure we don't break anything. But the problem will be users would have to assign manually a new _admin_ role to some users.

That's why we decided to proceed with the second option.

Reply:
It looks to me that immutable system roles are something we want to have anyway, _and_ it would make your life simpler, that's why I'm suggesting it here.

We are acting as if it was an unsolvable problem, but it's far from it. For example, we can have an immutable, editor-like system role that uses a fixed UUID as it's ID. It's not pretty, but it's a solution for older clusters. (Reserved role names would be even better, I wish we'd have prepared for that sooner.)

For the purposes of this design we don't need editor/auditor/access, we just need _a_ role that does the job.

Another alternative for this is the superuser approach on thread https://github.com/gravitational/teleport/pull/15706#discussion_r969887140.

Reply:
The problem is not with the name (we could even generate it) of the role but the need to assign some user to it: 

> But the problem will be users would have to assign manually a new _admin_ role to some users.